File: leetcode/572.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:

        def equals(l, r):
            if not l and not r:
                return True
            if not l or not r:
                return False
            return s.val == t.val and equals(l.left, r.left) and equals(l.right, r.right)
            

        return s and (equals(s, t) or self.isSubtree(s.left, t) or self.isSubtree(s.right, t))

File: leetcode/572.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // t is in s
    bool isSubtree(TreeNode* s, TreeNode* t) {
      if (!t && !s) return true;
      if (!s || !t) return false;

      if (check(s, t)) 
        return true;

      return isSubtree(s->left, t) || isSubtree(s->right, t);
    }

private:
    bool check(TreeNode* s, TreeNode* t) {
      if (!s && !t) return true;
      if (!s || !t) return false;
      return s -> val == t -> val && check(s->left, t->left) && check(s->right, t->right);
    }
};
